dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -221,249 +272,371 @@ namespace winrt::Microsoft::Windows::PushNotifications::implementation
+                }) };
+
+                auto notificationPlatform{ PushNotificationHelpers::GetNotificationPlatform() };
+                wil::unique_cotaskmem_string unpackagedAppUserModelId;
+                // Apps treated as unpackaged need to call RegisterFullTrustApplication and register with the LRP
+                THROW_IF_FAILED(notificationPlatform->RegisterFullTrustApplication(m_processName.get(), GUID_NULL, &unpackagedAppUserModelId));
 
+                // Register a sink for the application to receive foreground raw notifications
+                RegisterSinkHelper();
                 {
-                    auto lock{ s_activatorInfoLock.lock_shared() };
-                    THROW_HR_IF_MSG(E_INVALIDARG, s_comActivatorRegistration, "ComActivator already registered.");
+                    auto lock{ m_lock.lock_exclusive() };
+                    m_singletonForegroundRegistration = true;
+                    m_singletonBackgroundRegistration = true;
                 }
 
-                GetWaitHandleForArgs().create();
-                GetComServerClsid() = taskClsid;
-                
-                THROW_IF_FAILED(::CoRegisterClassObject(
-                    taskClsid,
-                    winrt::make<PushNotificationBackgroundTaskFactory>().get(),
-                    CLSCTX_LOCAL_SERVER,
-                    REGCLS_MULTIPLEUSE,
-                    &s_comActivatorRegistration));
-            }
+                // Register a COM object for the PushNotificationLongRunningProcess to CoCreate in background activation scenarios
+                if (AppModel::Identity::IsPackagedProcess())
+                {
+                    auto lock{ m_lock.lock_exclusive() };
+                    THROW_IF_FAILED(::CoRegisterClassObject(
+                        m_registeredClsid,
+                        winrt::make<PushNotificationBackgroundTaskFactory>().get(),
+                        CLSCTX_LOCAL_SERVER,
+                        REGCLS_MULTIPLEUSE,
+                        &m_comActivatorRegistration));
+                }
 
-            if (builder)
-            {
-                registeredTaskFromBuilder = builder.Register();
+                scopeExitToCleanRegistrations.release();
             }
-
-            scopeExitToCleanRegistrations.release();
-            auto lock{ s_activatorInfoLock.lock_exclusive() };
-            s_pushTriggerRegistration = registeredTaskFromBuilder;
-            PushNotificationTelemetry::ActivatorRegisteredByApi(S_OK, details.Activators());
         }
-
-        catch(...)
+        catch (...)
         {
-            PushNotificationTelemetry::ActivatorRegisteredByApi(wil::ResultFromCaughtException(),
-                details == nullptr ? PushNotificationRegistrationActivators::Undefined : details.Activators());
+            PushNotificationTelemetry::ActivatorRegisteredByApi(wil::ResultFromCaughtException());
             throw;
```

Comment:
Do we need to reset m_registering to false before throw?

Reply:
In general, the patern of throwing in a catch should be discouraged. 

Instead we should catch the hresult here and throw outside the try - catch